apptest.go

```diff
@@ -0,0 +1,192 @@
+)
+
+const (
+	deployedStatus = "deployed"
+	namespace      = "giantswarm"
+)
+
+// Config represents the configuration used to setup the apps.
+type Config struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+}
+
+// AppSetup implements the logic for managing the app setup.
+type AppSetup struct {
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+}
+
+// New creates a new configured app setup library.
+func New(config Config) (*AppSetup, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+
+	a := &AppSetup{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return a, nil
+}
+
+// InstallApps creates appcatalog and app CRs for use in automated tests
+// and ensures they are installed by our app platform.
+func (a *AppSetup) InstallApps(ctx context.Context, apps []App) error {
+	var err error
+
+	err = a.createAppCatalogs(ctx, apps)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = a.createApps(ctx, apps)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = a.waitForDeployedApps(ctx, apps)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (a *AppSetup) createAppCatalogs(ctx context.Context, apps []App) error {
+	var err error
+
+	for _, app := range apps {
+		a.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating %#q appcatalog cr", app.CatalogName))
+
+		appCatalogCR := &v1alpha1.AppCatalog{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: app.CatalogName,
+				Labels: map[string]string{
+					// Processed by app-operator-unique.
+					label.AppOperatorVersion: "0.0.0",
+				},
+			},
+			Spec: v1alpha1.AppCatalogSpec{
+				Description: app.CatalogName,
+				Title:       app.CatalogName,
+				Storage: v1alpha1.AppCatalogSpecStorage{
+					Type: "helm",
+					URL:  app.CatalogURL,
+				},
+			},
+		}
+		_, err = a.k8sClient.G8sClient().ApplicationV1alpha1().AppCatalogs().Create(ctx, appCatalogCR, metav1.CreateOptions{})
+		if apierrors.IsAlreadyExists(err) {
+			a.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("%#q appcatalog CR already exists", appCatalogCR.Name))
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		a.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created %#q appcatalog cr", app.CatalogName))
+	}
+
+	return nil
+}
+
+func (a *AppSetup) createApps(ctx context.Context, apps []App) error {
+	var err error
+
+	for _, app := range apps {
+		a.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating %#q app cr", app.Name))
+
+		appCR := &v1alpha1.App{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      app.Name,
+				Namespace: app.Namespace,
+				Labels: map[string]string{
+					// Processed by app-operator-unique.
+					label.AppOperatorVersion: "0.0.0",
```

Comment:
can we keep "0.0.0" as constant as well?

Reply:
Yes good point. Done.